src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/matchtoterm/TermFieldRewriter.java

```diff
@@ -210,19 +214,53 @@ public boolean isValidIdentifierForTerm(SQLIdentifierExpr expr) {
             );
     }
 
-    public boolean hasIdenticalMappings(TermFieldScope scope, Map<String, String> indexToType) {
+    private void checkMappingCompatibility(TermFieldScope scope, Map<String, String> indexToType) {
         if (scope.getMapper().isEmpty()) {
             throw new VerificationException("Unknown index " + indexToType.keySet());
         }
+        // Collect all FieldMappings into hash set and ignore index/type names. set.size > 1 means FieldMappings NOT unique,
+        // needs further compatibility check
+        Set<FieldMappings> mappingSet = curScope().getMapper().allMappings().stream().
+                                 flatMap(typeMappings -> typeMappings.allMappings().stream()).
+                                 collect(Collectors.toSet());
 
-        if (isMappingOfAllIndicesDifferent()) {
-            return false;
-        }
+        FieldMappings fieldMappings = null;
+
+        if (mappingSet.size() > 1) {
+            Map<String, Map<String, Object>> mergedMapping = new HashMap<>();
 
+            for (FieldMappings f : mappingSet) {
+                Map<String, Map<String, Object>> m = f.data();
+                m.forEach((k, v) -> getMergedMapping(k, v, mergedMapping));
+            }
+
+            fieldMappings = new FieldMappings(mergedMapping);
+        } else {
+            fieldMappings = curScope().getMapper().firstMapping().firstMapping();
+        }
         // We need finalMapping to lookup for rewriting
-        FieldMappings fieldMappings = curScope().getMapper().firstMapping().firstMapping();
         curScope().setFinalMapping(fieldMappings);
-        return true;
+    }
+
+    private void getMergedMapping(String fieldName, Map<String, Object> fieldMappingValue , Map<String, Map<String, Object>> mergedMapping){
+
+        if (!mergedMapping.containsKey(fieldName)) {
+            mergedMapping.put(fieldName, fieldMappingValue);
+        } else {
+            // check if types are same
+            if (!fieldMappingValue.equals(mergedMapping.get(fieldName))) {
+                // TODO: Merge mappings if they are compatible, like text and text/keyword to text/keyword.
+
+                String firstFieldType= new JSONObject(fieldMappingValue).toString().replaceAll("\"", "");
+                String secondFieldType = new JSONObject(mergedMapping.get(fieldName)).toString().replaceAll("\"", "");
+
+                String exceptionReason = String.format(Locale.ROOT,"Cannot use field [%s] " +
+                                "due to ambiguities being mapped as incompatible types: [%s] and [%s] ",
```

Comment:
message still not fixed

Reply:
I don't see any problem with the message. The message is in alignment with xpack exception. What is that you don't understand or seems incorrect to you. If you have good suggestion please provide it.

Reply:
> message still not fixed

Okay, we will use the followoing meassage as suggested `Different mappings are not allowed for the same field[%s]: found [%s] and [%s]